### PR TITLE
[pt-br] 5.8 file-system

### PR DESCRIPTION
**Pull request by thalesbruno**
First file translated (1/6):

Section 5.8 file-system
- [x] how-to-read-files-in-nodejs.md (@thalesbruno)
- [x]   how-to-search-files-and-directories-in-nodejs.md (@thalesbruno)
- [x]   how-to-store-local-config-data.md (@thalesbruno)
- [x]   how-to-use-the-path-module.md (@thalesbruno)
- [x]   how-to-write-files-in-nodejs.md (@thalesbruno)
- [x]   secutiry/introduction.md (@thalesbruno)

@glaucia86 @khaosdoctor

### CODE REVIEW

**Review by khaosdoctor — COMMENTED** [review]
File 1 seems ok

**Review by TiagoDanin — COMMENTED** [review]
File `how-to-read-files-in-nodejs.md` ok :+1:

**Comment by ghost** [comment]
@thalesbruno：I wonder whether you are going to translate all of the files in your list, because your title just has only one file translation...So if possible, please tell me and I'll add "on hold" to block and it will be released until all of your translations are finished?

**Comment by thalesbruno** [comment]
> @thalesbruno：I wonder whether you are going to translate all of the files in your list, because your title just has only one file translation...So if possible, please tell me and I'll add "on hold" to block and it will be released until all of your translations are finished?

Oh yes, I changed the title. So if you can add "on hold" I'll be grateful.

**Comment by thalesbruno** [comment]
job done @khaosdoctor @TiagoDanin @glaucia86

File: locale/pt-br/knowledge/file-system/how-to-search-files-and-directories-in-nodejs.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    fs.readdir(process.cwd(), function (err, files) {
+      if (err) {
+        console.log(err);
+        return;
+      }
+      console.log(files);
+    });
+
+
+Infelizmente, se você quuiser fazer uma lista recursiva dos arquivos, então as coisas rapidamente ficarão muito mais complicadas. Para evitar toda essa complexidade assustadora, esta é uma das circustâncias onde uma biblioteca feita por usuários salvar o dia. [Node-findit](https://github.com/substack/node-findit), pela SubStack, é um módulo auxiliar para facilitar a busca por arquivos.  Ele tem interfaces para permitir você trabalhar com callbacks, eventos, ou simplesmente da velha forma síncrona (o que na maioria das vezes não é uma boa ideia).
```
Comment (TiagoDanin) [inline locale/pt-br/knowledge/file-system/how-to-search-files-and-directories-in-nodejs.md]:
```suggestion
Infelizmente, se você quiser fazer uma lista recursiva dos arquivos, então as coisas rapidamente ficarão muito mais complicadas. Para evitar toda essa complexidade assustadora, esta é uma das circunstâncias onde uma biblioteca feita por usuários salvar o dia. [Node-findit](https://github.com/substack/node-findit), pela SubStack, é um módulo auxiliar para facilitar a busca por arquivos.  Ele tem interfaces para permitir você trabalhar com callbacks, eventos, ou simplesmente da velha forma síncrona (o que na maioria das vezes não é uma boa ideia).
```

File: locale/pt-br/knowledge/file-system/how-to-search-files-and-directories-in-nodejs.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+Para instalar `node-findit`, simplemente use npm:
```
Comment (TiagoDanin) [inline locale/pt-br/knowledge/file-system/how-to-search-files-and-directories-in-nodejs.md]:
```suggestion
Para instalar `node-findit`, simplesmente use npm:
```

File: locale/pt-br/knowledge/file-system/how-to-store-local-config-data.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+date: '2011-08-26T10:08:50.000Z'
+tags:
+  - conventions
+  - filesystem
+difficulty: 1
+layout: knowledge-post.hbs
+---
+
+
+Armazenar os dados de configuração da sua aplicação Node.js é muito simples - cada objeto no JavaScript pode ser facilmente rendererizado como [JSON](/what-is-json), que por sua vez é apenas dados em string que podem ser enviados ou salvos da maneira que você preferir.  A forma mais simples de fazer isso envolve os métodos built-in `JSON.parse()` e `JSON.stringify()`.
```
Comment (TiagoDanin) [inline locale/pt-br/knowledge/file-system/how-to-store-local-config-data.md]:
```suggestion
Armazenar os dados de configuração da sua aplicação Node.js é muito simples - cada objeto no JavaScript pode ser facilmente renderizado como [JSON](/what-is-json), que por sua vez é apenas dados em string que podem ser enviados ou salvos da maneira que você preferir.  A forma mais simples de fazer isso envolve os métodos built-in `JSON.parse()` e `JSON.stringify()`.
```

File: locale/pt-br/knowledge/file-system/how-to-store-local-config-data.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        console.dir(myObj);
+      }
+      catch (err) {
+        console.log('There has been an error parsing your JSON.')
+        console.log(err);
+      }
+
+NODE PRO TIP:  Mesmo que você não goste de usar `try/catch`, este é um lugar para usá-lo.  `JSON.parse` é um analisador JSON muito rígido, e erros são comuns - mais importante, contudo, `JSON.parse` usa a declaração `throw`  em vez de dar uma callback, então `try/catch` é o único jeito de se protejer contra os erros.
+
+Usar o método built-in `JSON` pode lhe levar longe, mas como como tantos outros problemas que você pode estar procurando resolver com Node.js, já há uma solução da comunidade que pode levar você muito além.  A solução, neste caso, é o `nconf`.  Escrito por Charlie Robbins, é um gerenciador de configurações para Node.js que suporta armazenamento em memória, armazenamento de arquivos local, bem como suporte a backend `redis`, provido por um módulo separado.
```
Comment (TiagoDanin) [inline locale/pt-br/knowledge/file-system/how-to-store-local-config-data.md]:
```suggestion
Usar o método built-in `JSON` pode lhe levar longe, mas como tantos outros problemas que você pode estar procurando resolver com Node.js, já há uma solução da comunidade que pode levar você muito além.  A solução, neste caso, é o `nconf`.  Escrito por Charlie Robbins, é um gerenciador de configurações para Node.js que suporta armazenamento em memória, armazenamento de arquivos local, bem como suporte a backend `redis`, provido por um módulo separado.
```

File: locale/pt-br/knowledge/file-system/security/introduction.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+title: Como eu posso proteger meu código?
+date: null
+tags:
+  - filesystem
+  - security
+difficulty: 3
+layout: knowledge-post.hbs
+---
+
+Às vezes, você pode querer deixar usuários lerem ou escreverem arquivos em seu servidor. Por exemplo, talvez você queira escrever um software de fórum sem usar um banco de dados de verdade. O problema é que você não quer que seus usuários possam modificar ou ler todo e qualquer arquivo em seu servidor, e por vezes há formas de contornar restrições que deveriam impedí-los. Continue lendo para ver como você pode proteger seu código contra atacantes malvados tentando bagunçar seus arquivos.
```
Comment (TiagoDanin) [inline locale/pt-br/knowledge/file-system/security/introduction.md]:
```suggestion
Às vezes, você pode querer deixar usuários lerem ou escreverem arquivos em seu servidor. Por exemplo, talvez você queira escrever um software de fórum sem usar um banco de dados de verdade. O problema é que você não quer que seus usuários possam modificar ou ler todo e qualquer arquivo em seu servidor, e por vezes há formas de contornar restrições que deveriam impedi-los. Continue lendo para ver como você pode proteger seu código contra atacantes malvados tentando bagunçar seus arquivos.
```

**Review by TiagoDanin — CHANGES_REQUESTED** [review]

File: locale/pt-br/knowledge/file-system/how-to-search-files-and-directories-in-nodejs.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
+---
+title: Como eu procuro arquivos e diretórios?
+date: '2011-08-26T10:08:50.000Z'
+tags:
+  - filesystem
+difficulty: 1
+layout: knowledge-post.hbs
+---
+
+Suponha que você queira listar todos os arquivos no seu diretório atual.  Uma abordagem é utilizar a função builtin `fs.readdir` [method](/how-do-i-read-files-in-node-js). Isto lhe retornará um array com todos os arquivos e diretórios do caminho especificado:
```
Comment (khaosdoctor) [inline locale/pt-br/knowledge/file-system/how-to-search-files-and-directories-in-nodejs.md]:
```suggestion
Suponha que você queira listar todos os arquivos no seu diretório atual.  Uma abordagem é utilizar a [função](/how-do-i-read-files-in-node-js) builtin `fs.readdir`. Isto lhe retornará um array com todos os arquivos e diretórios do caminho especificado:
```

File: locale/pt-br/knowledge/file-system/how-to-store-local-config-data.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+      try {
+        myObj = JSON.parse(data);
+        console.dir(myObj);
+      }
+      catch (err) {
+        console.log('There has been an error parsing your JSON.')
+        console.log(err);
+      }
+
+NODE PRO TIP:  Mesmo que você não goste de usar `try/catch`, este é um lugar para usá-lo.  `JSON.parse` é um analisador JSON muito rígido, e erros são comuns - mais importante, contudo, `JSON.parse` usa a declaração `throw`  em vez de dar uma callback, então `try/catch` é o único jeito de se protejer contra os erros.
```
Comment (khaosdoctor) [inline locale/pt-br/knowledge/file-system/how-to-store-local-config-data.md]:
```suggestion
NODE PRO TIP:  Mesmo que você não goste de usar `try/catch`, este é um lugar para usá-lo.  `JSON.parse` é um analisador JSON muito rígido, e erros são comuns - mais importante, contudo, `JSON.parse` usa a declaração `throw`  em vez de dar uma callback, então `try/catch` é o único jeito de se proteger contra os erros.
```

File: locale/pt-br/knowledge/file-system/how-to-store-local-config-data.md (hunk 4 above)
Comment (khaosdoctor) [inline locale/pt-br/knowledge/file-system/how-to-store-local-config-data.md]:
```suggestion
Usar o método built-in `JSON` pode te levar longe, mas, como tantos outros problemas que você pode estar procurando resolver com Node.js, já há uma solução da comunidade que pode levar você muito além.  A solução, neste caso, é o `nconf`.  Escrito por Charlie Robbins, é um gerenciador de configurações para Node.js que suporta armazenamento em memória, armazenamento de arquivos local, bem como suporte a backend `redis`, provido por um módulo separado.
```

**Review by TiagoDanin — APPROVED** [review]
:100: LGTM

File: locale/pt-br/knowledge/file-system/how-to-use-the-path-module.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    > path.basename(a, path.extname(a))
+    'c'
+
+Por último, o módulo `path` fornece métodos para checar se um determinado caminho existe ou não: `exists` e `existsSync` Os dois recebem o caminho de um arquivo como primeiro parâmetro. 
+
+`exists` recebe uma callback como seu segundo parâmetro, a qual é retornado um booleano representando a existência do arquivo. 
+
+`existsSync`, por outro lado, checa o caminho recebido de forma síncrona.  No Node.js, você normalmente irá querer usar as funções assíncronas para entrada/saída no sistema de arquivos - as versões síncronas bloquearão todo o seu processo até que esteja finalizado. 
+
+Bloqueio nem sempre é uma coisa ruim.  Checar a existência de um arquivo de configuração essencial de forma síncrona faz sentido, por exemplo - não interessa muito se seu processo está bloqueado por algo que ele não pode executar sem!  Por outro lado, entretanto, em um servidor HTTP muito ativo, qualquer entrada/saída de arquivos por requisição **DEVE** ser assíncrona, senão você responderá as requisições uma por uma. Veja o artigo em [asynchronous operations](/how-to-write-asynchronous-code) para mais detalhes.
```
Comment (khaosdoctor) [inline locale/pt-br/knowledge/file-system/how-to-use-the-path-module.md]:
```suggestion
Bloqueio nem sempre é uma coisa ruim.  Checar a existência de um arquivo de configuração essencial de forma síncrona faz sentido, por exemplo - não interessa muito se seu processo está bloqueado por algo que ele não pode viver sem!  Por outro lado, entretanto, em um servidor HTTP muito ativo, qualquer entrada/saída de arquivos por requisição **DEVE** ser assíncrona, senão você responderá as requisições uma por uma. Veja o artigo em [operações assíncronas](/how-to-write-asynchronous-code) para mais detalhes.
```

File: locale/pt-br/knowledge/file-system/how-to-write-files-in-nodejs.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+Então, se nós quisermos escrever "Hello World" em `helloworld.txt`:
+
+    fs = require('fs');
+    fs.writeFile('helloworld.txt', 'Hello World!', function (err) {
+      if (err) return console.log(err);
+      console.log('Hello World > helloworld.txt');
+    });
+
+    [contents of helloworld.txt]:
```
Comment (khaosdoctor) [inline locale/pt-br/knowledge/file-system/how-to-write-files-in-nodejs.md]:
```suggestion
    [conteúdo de helloworld.txt]:
```

File: locale/pt-br/knowledge/file-system/security/introduction.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+=================
+Poison null bytes é um modo de enganar o seu código para que ele veja outro nome de arquivo em vez do que realmente está sendo aberto. Isto pode, em muitos casos, ser usado para contornar proteções a directory traversal, para enganar servidores a entregar arquivos com tipos errados e para contornar restrições nos nomes de arquivos que podem ser usados. [Aqui está uma descrição mais detalhada.](http://groups.google.com/group/nodejs/browse_thread/thread/51f66075e249d767/85f647474b564fde) Sempre use código como este quando estiver acessando arquivos com nomes fornecidos pelo usuário:
+
+    if (filename.indexOf('\0') !== -1) {
+      return respond('That was evil.');
+    }
+
+Whitelisting
+============
+Você sem sempre será capaz de usar whitelisting, mas se você for, use-a - é muito fácil de implementar e difícil de dar errado. Por exemplo, se você souber que todos os nomes de arquivos são strings alfanuméricas em caixa baixa:
```
Comment (khaosdoctor) [inline locale/pt-br/knowledge/file-system/security/introduction.md]:
```suggestion
Você nem sempre será capaz de usar whitelisting, mas se for, é muito fácil de implementar e difícil de dar errado. Por exemplo, se você souber que todos os nomes de arquivos são strings alfanuméricas em caixa baixa:
```

File: locale/pt-br/knowledge/file-system/security/introduction.md [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+    if (!/^[a-z0-9]+$/.test(filename)) {
+      return respond('illegal character');
+    }
+
+Contudo, note que whitelisting sozinha não será mais suficiente assim que  você permitir pontos e barras - pessoas poderiam entrar com coisas como `../../etc/passwd` a fim de obter arquivos de fora da pasta permitida.
```
Comment (khaosdoctor) [inline locale/pt-br/knowledge/file-system/security/introduction.md]:
```suggestion
Contudo, note que whitelisting sozinha não será suficiente assim que você permitir pontos e barras - pessoas poderiam entrar com coisas como `../../etc/passwd` a fim de obter arquivos de fora da pasta permitida.
```

File: locale/pt-br/knowledge/file-system/security/introduction.md [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+Prevenindo Directory Traversal
+==============================
+Directory traversal significa que um atacante tenta acessar arquivos fora do diretório que você quer permitir que ele acesse. Você pode prevenir isto usando o módulo "path" nativo do nodes. **Não reimplementar as coisas no módulo path por conta própria** - por exemplo, quando alguém executa seu código em um servidor windows, não manipular barras invertidas como barras normais permitrá que atacantes façam directory traversal.
```
Comment (khaosdoctor) [inline locale/pt-br/knowledge/file-system/security/introduction.md]:
```suggestion
Directory traversal significa que um atacante tenta acessar arquivos fora do diretório que você quer permitir que ele acesse. Você pode prevenir isto usando o módulo "path" nativo do Node.js. **Não re-implementar as coisas no módulo path por conta própria** - por exemplo, quando alguém executa seu código em um servidor windows, não manipular barras invertidas como barras normais permitirá que atacantes façam directory traversal.
```

**Review by khaosdoctor — CHANGES_REQUESTED** [review]
A few minor changes

File: locale/pt-br/knowledge/file-system/security/introduction.md (hunk 12 above)
Comment (thalesbruno) [reply locale/pt-br/knowledge/file-system/security/introduction.md]: Thanks. I just disagree about "reimplementar", there's no hyphen.
https://ciberduvidas.iscte-iul.pt/consultorio/perguntas/a-grafia-de-palavras-com-o-prefixo-re-rerromanizacao/33575
https://www.soportugues.com.br/secoes/FAQresposta.php?id=126

**Review by khaosdoctor — APPROVED** [review]

File: locale/pt-br/knowledge/file-system/security/introduction.md (hunk 12 above)
Comment (khaosdoctor) [reply locale/pt-br/knowledge/file-system/security/introduction.md]: Yes you are correct!

**Review by khaosdoctor — APPROVED** [review]

**Review by khaosdoctor — APPROVED** [review]

**Comment by khaosdoctor** [comment]
I think that's it @MaledongGit

**Pull request merged**